### PR TITLE
Updates tolerations when workbench starts

### PR DESCRIPTION
**Pull request by andrewballantyne**
Closes: #903 

## Description

DS Projects -- Updates tolerations as the Workbench starts (via the toggle). Edit should be unaffected.

## How Has This Been Tested?

1. Create a Workbench with no toleration setting configured (see Settings => Cluster settings)
2. Enable the toleration settings
3. Turn off/on the Workbench -- see the change (added)
4. Change the toleration setting value
5. Turn off/on the Workbench -- see the change (updated)
6. Disable the toleration settings
7. Turn off/on the Workbench -- see the change (removed)

Note: It should be noted we do not update the config immediately after saving. There is a notification update in #939 that covers this scenario. In dev mode refresh or wait for the config to be returned from the server before testing a turn-on state for the Workbench.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: frontend/src/utilities/tolerations.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  );
+
+  const hasTolerations = !!tolerations && tolerations.length > 0;
+  const doNothing = settings.length === 0 && !hasTolerations;
+  const tolerationsRemoved = settings.length === 0 && hasTolerations;
+  const updateTolerations = settings.length > 0 && hasTolerations;
+
+  let type: TolerationChanges['type'];
+  if (tolerationsRemoved) {
+    type = 'remove';
```
Comment (DaoDaoNoCode) [inline frontend/src/utilities/tolerations.ts:56]: If the type is `remove`, the patch will be
```
{
  op: 'remove',
  path: '/spec/template/spec/tolerations',
  value: [],
}
```
Is it valid here if the op is `remove` while we still set the value?

Comment (andrewballantyne) [reply frontend/src/utilities/tolerations.ts:56]: A fair point -- I'll fix it during rebase.

File: frontend/src/utilities/tolerations.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  const settings = determineTolerations(
+    hasGPU,
+    dashboardConfig.spec.notebookController?.notebookTolerationSettings,
+  );
+
+  const hasTolerations = !!tolerations && tolerations.length > 0;
+  const doNothing = settings.length === 0 && !hasTolerations;
+  const tolerationsRemoved = settings.length === 0 && hasTolerations;
+  const updateTolerations = settings.length > 0 && hasTolerations;
```
[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/src/utilities/tolerations.ts:53]: Yeah -- this is a fair point -- I contemplated fighting over the diff to make sure -- but it's a patch not an update. So our last computed state is the right state for it to be in (as far as the UI is concerned without refetching the world to determine that) -- if the server changed briefly before the patch and for some reason this was impacted, I don't really want to worry about that overlap and diff calculation.

It was a soft choice -- I'm not really set in my thinking here -- but the patch is cheap and the API call is happening anyways to enable the Workbench. 🤷 

Do you feel strongly one way or another? I could do the diff (and additional fetching if needed) if you think it adds notable value.

[the remark this answers is not in the document]
Comment (Gkrumbach07) [reply frontend/src/utilities/tolerations.ts:53]: The operation isn't visible to the user so as long as the api call is low cost then we can keep it how it is. I would only want to go change it if the user could see it. 

LGTM

**Comment by andrewballantyne** [comment]
Added lgtm based on [Gage's comment](https://github.com/opendatahub-io/odh-dashboard/pull/938#discussion_r1103049631)

**Pull request merged**